src/node_contextify.cc

```diff
@@ -1483,50 +1486,110 @@ static MaybeLocal<Function> CompileFunctionForCJSLoader(Environment* env,
+// about which file or which package.json to update.
+const char* require_esm_warning =
+    "To load an ES module, set \"type\": \"module\" in the package.json or use "
+    "the .mjs extension.";
+
+static bool ShouldRetryAsESM(Realm* realm,
+                             Local<String> message,
+                             Local<String> code,
+                             Local<String> resource_name);
+
 static void CompileFunctionForCJSLoader(
     const FunctionCallbackInfo<Value>& args) {
   CHECK(args[0]->IsString());
   CHECK(args[1]->IsString());
+  CHECK(args[2]->IsBoolean());
+  CHECK(args[3]->IsBoolean());
   Local<String> code = args[0].As<String>();
   Local<String> filename = args[1].As<String>();
+  bool is_main = args[2].As<Boolean>()->Value();
+  bool should_detect_module = args[3].As<Boolean>()->Value();
+
   Isolate* isolate = args.GetIsolate();
   Local<Context> context = isolate->GetCurrentContext();
-  Environment* env = Environment::GetCurrent(context);
+  Realm* realm = Realm::GetCurrent(context);
+  Environment* env = realm->env();
 
   bool cache_rejected = false;
   Local<Function> fn;
+  Local<Value> cjs_exception;
+  Local<Message> cjs_message;
+
   {
+    ShouldNotAbortOnUncaughtScope no_abort_scope(realm->env());
     TryCatchScope try_catch(env);
     if (!CompileFunctionForCJSLoader(
              env, context, code, filename, &cache_rejected)
              .ToLocal(&fn)) {
       CHECK(try_catch.HasCaught());
       CHECK(!try_catch.HasTerminated());
-      errors::DecorateErrorStack(env, try_catch);
-      try_catch.ReThrow();
+      cjs_exception = try_catch.Exception();
+      cjs_message = try_catch.Message();
+      errors::DecorateErrorStack(env, cjs_exception, cjs_message);
+    }
+  }
+
+  bool can_parse_as_esm = false;
+  if (!cjs_exception.IsEmpty()) {
+    // Use the URL to match what would be used in the origin if it's going to
+    // be reparsed as ESM.
+    Utf8Value filename_utf8(isolate, filename);
+    std::string url = url::FromFilePath(filename_utf8.ToStringView());
+    Local<String> url_value;
+    if (!String::NewFromUtf8(isolate, url.c_str()).ToLocal(&url_value)) {
+      return;
+    }
+    can_parse_as_esm =
+        ShouldRetryAsESM(realm, cjs_message->Get(), code, url_value);
+    if (!can_parse_as_esm) {
+      // The syntax error is not related to ESM, throw the original error.
+      isolate->ThrowException(cjs_exception);
+      return;
+    }
+
+    if (!should_detect_module) {
+      bool should_throw = true;
+      if (!warned_about_require_esm) {
+        // This needs to call process.emit('warning') in JS which can throw if
+        // the user listener throws. In that case, don't try to throw the syntax
+        // error.
+        should_throw =
+            ProcessEmitWarningSync(env, require_esm_warning).IsJust();
+      }
+      if (should_throw) {
+        isolate->ThrowException(cjs_exception);
+      }
```

Comment:
NIT: 
```suggestion
      // ProcessEmitWarningSync needs to call process.emit('warning') in JS which can throw if
      // the user listener throws. In that case, don't try to throw the syntax error.
      if (!warned_about_require_esm && ProcessEmitWarningSync(env, require_esm_warning).IsJust()) {
        isolate->ThrowException(cjs_exception);
      }
```
Just a shortening which might be easier to read

Reply:
The suggestion works differently: that means if it’s warned before, no more CJS parse error will ever be thrown again.

Reply:
```suggestion
      // ProcessEmitWarningSync needs to call process.emit('warning') in JS which can throw if
      // the user listener throws. In that case, don't try to throw the syntax error.
      if (warned_about_require_esm || ProcessEmitWarningSync(env, require_esm_warning).IsJust()) {
        isolate->ThrowException(cjs_exception);
      }
```
I mixed a bit there, of course. :/ This would be the equivalent shortening.

Reply:
I actually find this harder to reason about…